dev-tools/harness.js

```diff
@@ -0,0 +1,229 @@
+			log();
+			throw err;
+		});
+	}
+
+	async runGame(format, battleStream) {
+		const streams = BattleStreams.getPlayerStreams(battleStream);
+		const spec = {formatid: format, seed: this.prng.seed};
+		const p1spec = this.getPlayerSpec("Bot 1", this.p1options);
+		const p2spec = this.getPlayerSpec("Bot 2", this.p2options);
+
+		const p1 = new RandomPlayerAI(
+			streams.p1, Object.assign({seed: this.newSeed()}, this.p1options)).start();
+		const p2 = new RandomPlayerAI(
+			streams.p2, Object.assign({seed: this.newSeed()}, this.p2options)).start();
+
+		streams.omniscient.write(`>start ${JSON.stringify(spec)}\n` +
+			`>player p1 ${JSON.stringify(p1spec)}\n` +
+			`>player p2 ${JSON.stringify(p2spec)}`);
+
+		let chunk;
+		while ((chunk = await Promise.race([streams.omniscient.read(), p1, p2]))) {
+			if (this.output) console.log(chunk);
+		}
+	}
+
+	// Same as PRNG#generatedSeed, only deterministic.
+	// NOTE: advances this.prng's seed by 4.
+	newSeed() {
+		return [
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+		];
+	}
+
+	getPlayerSpec(name, options) {
+		if (options.team) return {name, team: options.team};
+		return {name, seed: this.newSeed()};
+	}
+}
+
+module.exports = Runner;
+
+const FORMATS = [
+	'gen7randombattle', 'gen7randomdoublesbattle', 'gen7battlefactory', 'gen6randombattle', 'gen6battlefactory',
+	'gen5randombattle', 'gen4randombattle', 'gen3randombattle', 'gen2randombattle', 'gen1randombattle',
+];
+
+class MultiRunner {
+	constructor(options) {
+		this.options = Object.assign({}, options);
+
+		this.totalGames = options.totalGames;
+
+		this.prng = (options.prng && !Array.isArray(options.prng)) ?
+			options.prng : new PRNG(options.prng);
+		this.options.prng = this.prng;
+
+		this.format = options.format;
+		this.cycle = !!options.cycle;
+		this.all = !!options.all;
+
+		this.async = !!options.async;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let format, lastFormat;
+		let failures = 0;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				// If we ran async, we need to now wait for each game and determine its status.
+				// NOTE: Promise.all doesn't help us here because it will resolve when the first
+				// rejection occurs, we need to wait for *all* the rejections.
+				if (this.async) for (const game of games) await game;
+				games = [];
+			}
+
+			const seed = this.prng.seed;
+			const game = new Runner(Object.assign({format}, this.options)).run().catch(err => {
+				failures++;
+				console.error(
+					`Run \`node dev-tools/harness 1 --format=${format} --seed=${seed.join()}\` ` +
+					`to debug (optionally with \`--output\` and/or \`--input\` for more info):\n`, err);
+			});
+
+			if (!this.async) await game;
+			games.push(game);
+			lastFormat = format;
+		}
```

Comment:
After removing the implementations of ``--async`` and ``--all``, I tried to rewrite this loop in all sorts of ways, but I couldn't convince V8 to GC the intermediate ``Promise`` and ``TickObject`` instances this creates...

An implementation that uses good old callbacks doesn't leak though :)
https://hastebin.com/odoguzecuw.js

Reply:
FYI: When I get the code functionally correct I'm going to dig into the memory issues, but I really appreciate that you spent the time to pinpointed the reason being promises. However, I dont think a callback based implementation is the way to go unless absolutely necessary, because `await` and `Promise` are idiomatic and if using them with the PS APIs causes issues then we probably need to fix the APIs

Reply:
```
[20:29:07] @Slayer95: pre, fixed the memory leak by doing await new Promise(resolve => {setImmediate(resolve)}) every N games run
[20:30:14] @Slayer95: as it turns out, there is not a single setImmediate in the BattleStream implementation, so it's in fact a sync stream
[20:30:50] @Slayer95: now, exactly how that impacts the GC logic isn't very clear
[20:30:56] @pre: i think thats a `BattleStream` or `Steams` issue then?
[20:31:01] @pre: but thanks for figuring that out
[20:31:15] @Slayer95: there are very few articles discussing memory leaks from looping async
[20:31:19] @pre: i was literally just commenting on your callback implementation from yesterday
[20:31:42] @Slayer95: yes, a fix in BattleStream / Streams should be feasible
```